### ISSUE TITLE
Possible Memory Leak in Reverb Implementations

### ISSUE THREAD

**Issue by btfranklin** [AudioKit/AudioKit#2245]
I have a relatively complex application which builds songs in memory dynamically using various pieces of the AKSequencer. In doing this, each song builds up its own signal chain with channels and effects. When songs are changed, everything gets torn down and rebuilt. This has been working really wonderfully, and everything is clean.

However, today I noticed that as I changed songs many times, the memory usage of my app in XCode was increasing each time. Using the Instruments tool and the leak identifier built into that, I was able to isolate what I believe may be a memory leak in AudioKit itself, specifically in the Reverb implementations.

I have attached a couple of images from the Instruments tool to show what I'm seeing. I have narrowed them down to only show the relevant parts.

<img width="929" alt="Screen Shot 2020-09-03 at 9 33 47 PM" src="https://user-images.githubusercontent.com/23176608/92200364-8544b400-ee2e-11ea-92d8-98f3a7802882.png">
<img width="924" alt="Screen Shot 2020-09-03 at 9 34 37 PM" src="https://user-images.githubusercontent.com/23176608/92200371-8aa1fe80-ee2e-11ea-84d4-5d05eb846ff7.png">

"DrumsTrackDefinition" is my own code. I have included the second image to show that I'm seeing this in three different Reverbs, with the AKZitaReverb seeming to be the worst offender w/r/t amount of memory leaking.

Here is what the "createRoute" method shown there from my code does. As you will see, it is straightforward:

```
    func createRoute(from signalNode: AKNode, to mixerNode: AKMixer) -> [AKNode] {
        
        var effectNodes = [AKNode]()
        
        let reverb: AKNode
          
        switch genotype.reverbImplementation {
        case .chowning:
            reverb = AKChowningReverb(signalNode)
        case .costello:
            reverb = AKCostelloReverb(signalNode)
        case .appleReverb:
            reverb = AKReverb(signalNode,
                              dryWetMix: 1.0)
        case .appleReverb2:
            reverb = AKReverb2(signalNode)
        case .zita:
            reverb = AKZitaReverb(signalNode,
                                  dryWetMix: 1.0)
        }
        effectNodes.append(reverb)
        
        let dryWetMixer = AKDryWetMixer(signalNode, reverb, balance: genotype.reverbMix)
        effectNodes.append(dryWetMixer)
        
        let localMixer = AKMixer(dryWetMixer)
        localMixer.volume = Volume(userValue: 0.7).mixerValue
        effectNodes.append(localMixer)
        
        localMixer >>> mixerNode

        return effectNodes
    }
```

The reason I do it this way is so that I can maintain a reference to the "effectNodes" elsewhere for later deallocation. So all that's happening here from my code's side is just instantiating these nodes, and they are being identified as memory leaks. (FTR, I use this exact same technique in many other places in the same product, and there are no leaks identified, but I only use the Reverb nodes on the Drums, so that is why I think it's being identified only here.) I don't know exactly how Instruments is able to tell this is a memory leak, but I assume it must be somehow noticing that no reference is being kept to the allocated memory, which means it cannot be deallocated later? I don't know.

Later, I do cleanup of the nodes like this, but it is **not** deallocating this memory. Hence the leak.

```
    deinit {
        for node in nodes {
            node.disconnectOutput()
            node.detach()
        }
    }
````

I will be happy to provide you any additional information that you might need or want to help debug this. Please let me know.

Thank you!

**Comment by aure** [comment]
Thanks for this, we should be able to track this down.

**Comment by aure** [comment]
I see you allow a choice of 5 reverbs but only see leaks in 3. Which 3? I'm assuming Costello and Chowning? For your other effects that do not show a leak, can you let me know which ones and which do not (in order of offensiveness is optional).  That will help me track down the problem. Thanks!

**Comment by btfranklin** [comment]
Hi! If you take a look at the second screenshot (the much smaller one), I collapsed the individual Reverb implementations to occupy only one line each. By far the worst offender appears to be Zita based on that. However, I did not do a comprehensive analysis to make sure that the system used every type of reverb, and the leak could have resulted from the same reverb being chosen multiple times. I can do a test with a much larger number of song generations to try to give you some more detailed results.

**Comment by aure** [comment]
Okay, thanks that makes sense.

**Comment by btfranklin** [comment]
Okay, I just did a test that instantiated 32 songs. Keeping in mind that they're generated totally randomly (more or less), the only three reverb implementations I saw come up were: Zita, Chowning, Costello. They were in that order of severity, and the results almost perfectly matched what was in that original screenshot up there. So I'd definitely start with Zita since its leak is approximately 3 times as large as Chowning's.

**Incidentally**, I just happened to notice that there is one other leak happening, which is _very, very_ small but possibly also worth noting just for completeness. I've attached it to this comment in case you want to follow up on it as well, but I'd place it as being much less urgent.
<img width="924" alt="Screen Shot 2020-09-06 at 11 01 54 PM" src="https://user-images.githubusercontent.com/23176608/92353385-09987080-f095-11ea-80f0-bd4b51eab2f8.png">

**Comment by PaulBatchelor** [comment]
It would seem like the soundpipe destroy functions are not being called. `sp_create` needs a corresponding `sp_destroy`. `sp_zitarev_init` and `sp_zitarev_create` pair with `sp_zitarev_destroy`. And so on. Is there any way to check and make sure those are being called? And if they are being called, are they being called the right number of times?

**Comment by PaulBatchelor** [comment]
FWIW, these do not leak at the soundpipe level for me. No leaks show up for me when I run this program in Valgrind:

```c
#include "soundpipe.h"

int main(int argc, char *argv[])
{
    sp_data *sp;
    sp_zitarev *zrev;
    sp_revsc *revsc;
    sp_jcrev *jrev;

    sp_create(&sp);

    sp_zitarev_create(&zrev);
    sp_zitarev_init(sp, zrev);

    sp_revsc_create(&revsc);
    sp_revsc_init(sp, revsc);

    sp_jcrev_create(&jrev);
    sp_jcrev_init(sp, jrev);

    sp_jcrev_destroy(&jrev);
    sp_zitarev_destroy(&zrev);
    sp_revsc_destroy(&revsc);
    sp_destroy(&sp);
    return 0;
}
```

**Comment by btfranklin** [comment]
From what I can understand by looking at the code, it looks like the corresponding 'destroy' functions that you're referring to get called in the deinit methods, such as AKZitaReverbDSP.mm. It seems like it would be fruitful to ensure that those actually get called.

**Comment by aure** [comment]
I made myself a little example project and proved that the deinit was being called, but memory was still leaking. Paul might be onto something in that its possible init is being called twice and deinit only once.

**Comment by aure** [comment]
In my test project init is called once and deinit is called once.

**Comment by aure** [comment]
I should say my project does also show the memory usage going up, confirming the leak.

**Comment by btfranklin** [comment]
So, it's probably worth pointing out that Instruments in Xcode identifies this as a leak *immediately*. It does not call it a leak after an attempt to dealloc the memory. Basically as soon as it's allocated, it's called a leak. As I said above, I'm not entirely sure how Instruments identifies leaks, but my guess in this case is that perhaps it notices that a piece of memory is allocated and there's no pointer that accurately tracks that memory for later deallocation. Could it be that the pointer used during the destroy process is not correct? Sorry, it's been many years since I did any C coding so I'm a bit rusty.

**Comment by btfranklin** [comment]
What I'm getting at is that it's noteworthy that Instruments can tell it's a leak as soon as the allocation is finished. That implies that the mistake is being made during allocation, right?

**Comment by aure** [comment]
Out of curiosity, is it possible this is only happening when connected to Xcode in debug mode? I put a button on my interface to destroy and recreate the reverbs and when I disconnected from Xcode the memory use seemed much more stable. Maybe still a little bit increasing, but not nearly as much.

**Comment by btfranklin** [comment]
I'm honestly not sure. When I'm seeing this, I have actually transferred the attachment from XCode to Instruments, so it restarts the app attached only to Instruments. As far as I know, I don't even have an XCode debug session running at all when I'm seeing this. That said, it could be something about how Instruments is watching the data that's keeping it around...? That seems kind of weird/unlikely but I guess anything is possible.

**Comment by aure** [comment]
@btfranklin Do you think you could try moving to AudioKit v5-develop and using a mixer and it's addInput and removeInput features to do your dynamic stuff and see if this problem is still there? Otherwise I might postpone this issue until v5 is out.

**Comment by btfranklin** [comment]
I'll give it a shot. I've built v5-develop locally and I'm currently working to integrate it into my project. It didn't pop in completely seamlessly, though...There seem to be some changes such as AudioKit.start(), AudioKit.output, and AudioKit.midi no longer being present. It's not immediately obvious where I can look to see updated docs or guidance, so for the moment I'll try to get through via experimentation.

**Comment by btfranklin** [comment]
Okay, with no other changes beyond what I just described for initializing/starting up the Engine itself, I'm still seeing the exact same results in Instruments. Zita, Chowning, and Costello, with Zita being by far the worst leak in terms of size.

As far as the addInput and removeInput features you're referring to, I'm not 100% sure what I need to do differently to do what you're asking. I currently use the >>> operator for most of these mixer connections in my code. Here's an example of how I create a route.
```    
func createRoute(from signalNode: AKNode, to mixerNode: AKMixer) -> [AKNode] {
        
        var effectNodes = [AKNode]()
        
        let compressor = AKCompressor(signalNode)
        effectNodes.append(compressor)
        
        let localMixer = AKMixer(compressor)
        localMixer.volume = Volume(userValue: 0.5).mixerValue
        effectNodes.append(localMixer)
        
        localMixer >>> mixerNode
        
        return effectNodes
    }
```

The "localMixer" is so I can adjust the volume of that route independently. Should I be avoiding using the >>> method for attachment?

**Comment by btfranklin** [comment]
Follow up: Just to play it safe, I added "mainMixer.disconnectInput()" to my disconnection procedure, which is where I usually dispose of the entire signal chain (with the exception of the main mixer and EQ band filters, which stay in place as everything else gets created and destroyed). If I understand it correctly, this would remove *all* input connections, including this connection from the local mixers. This does not seem to have had any effect, and I'm still seeing the same leak.

**Comment by aure** [comment]
Any chance that you could confirm this still happens in AudioKit 5. Just not sure what to do about this, so I will close it for now.

**Comment by btfranklin** [comment]
Yes, it still happens with AudioKit 5 as viewed through XCode and Instruments. I'll do some tests in a bit to see if it happens when running in "production" mode.

**Comment by inhumanscale** [comment]
i'm running into this as well.  I haven't migrated to ak5 yet but it sounds like it may be the same issue.
